From 26de1f81b4ec9ef8222c5e5d5d5d1d3d24b9e726 Mon Sep 17 00:00:00 2001
From: rramasamy <rramasamy@maxlinear.com>
Date: Tue, 10 Jun 2025 18:25:50 +0800
Subject: [PATCH] bwl: whm: send channel preference report to controller

As per the Easymesh Standard, if a Multi-AP-Agent detects a change in DFS
state of a channel it should send the unsolicited Channel Preference report
to the Controller.

As per the Airties requirements, whenever the Agent detects radar it should
send Channel Preference Report message indicating that radar has been found
both in Channel Preference TLV and CAC Status TLV, which has been done
earlier for RDK Platform using the CSA_Finished event.

use the existing flow for this requirement to send the report to Controller
whenever CSA_Finished event with reason "RADAR" is received.

use RadarTriggeredDfsChannels parameter in ChannelMgt object to get the Radar
triggered channels for filling Channel Preference and CAC Status TLV.

Closes PPM-3278

Signed-off-by: rramasamy <rramasamy@maxlinear.com>
---
 common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp   | 12 ++++++++
 common/beerocks/bwl/whm/base_wlan_hal_whm.cpp | 30 +++++++++++--------
 2 files changed, 30 insertions(+), 12 deletions(-)

diff --git a/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp b/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp
index 0ba56bd7b6..0f0cf1a6da 100644
--- a/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp
+++ b/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp
@@ -41,6 +41,8 @@ static ap_wlan_hal::Event wpaCtrl_to_bwl_event(const std::string &opcode)
         return ap_wlan_hal::Event::DFS_CAC_Completed;
     } else if (opcode == "DFS-NOP-FINISHED") {
         return ap_wlan_hal::Event::DFS_NOP_Finished;
+    } else if (opcode == "AP-CSA-FINISHED") {
+        return ap_wlan_hal::Event::CSA_Finished;
     } else if (opcode == "CTRL-EVENT-EAP-FAILURE") {
         return ap_wlan_hal::Event::WPA_Event_EAP_Failure;
     } else if (opcode == "CTRL-EVENT-EAP-FAILURE2") {
@@ -1580,6 +1582,16 @@ bool ap_wlan_hal_whm::process_wpa_ctrl_event(const beerocks::wbapi::AmbiorixVari
         event_queue_push(Event::DFS_NOP_Finished, msg_buff);
         break;
     }
+    case Event::CSA_Finished: {
+        const std::string reason  = parsed_obj["reason"];
+        ChanSwReason chanSwReason = ChanSwReason::Unknown;
+        if (reason == "RADAR") {
+            chanSwReason = ChanSwReason::Radar;
+        }
+        m_radio_info.last_csa_sw_reason = chanSwReason;
+        event_queue_push(event);
+        break;
+    }
     case Event::WPA_Event_EAP_Failure:
     case Event::WPA_Event_EAP_Failure2:
     case Event::WPA_Event_EAP_Timeout_Failure:
diff --git a/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp b/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
index 243c0f8046..3a098418a9 100644
--- a/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
+++ b/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
@@ -444,24 +444,30 @@ bool base_wlan_hal_whm::refresh_radio_info()
     radio->read_child(m_radio_info.channel, "Channel");
     m_radio_info.is_dfs_channel = son::wireless_utils::is_dfs_channel(m_radio_info.channel);
 
+    std::unordered_set<std::string> cleared_channels_set;
+    std::unordered_set<std::string> radar_triggered_channels_set;
+
+    std::string channel_mgt_path = m_radio_path + "ChannelMgt.";
+    auto channel_mgt_obj         = m_ambiorix_cl.get_object(channel_mgt_path);
+    if (!channel_mgt_obj) {
+        LOG(ERROR) << "failed to get radio Stats object " << channel_mgt_path;
+        return true;
+    }
+    channel_mgt_obj->read_child(s_val, "ClearedDfsChannels");
+    auto cleared_channels_vec = beerocks::string_utils::str_split(s_val, ',');
+    cleared_channels_set.insert(cleared_channels_vec.cbegin(), cleared_channels_vec.cend());
+
+    channel_mgt_obj->read_child(s_val, "RadarTriggeredDfsChannels");
+    auto radar_triggered_channels_vec = beerocks::string_utils::str_split(s_val, ',');
+    radar_triggered_channels_set.insert(radar_triggered_channels_vec.cbegin(),
+                                        radar_triggered_channels_vec.cend());
+
     if (radio->read_child(s_val, "PossibleChannels")) {
         auto channels_vec = beerocks::string_utils::str_split(s_val, ',');
         for (auto &chan_str : channels_vec) {
             uint32_t chanNum   = beerocks::string_utils::stoi(chan_str);
             auto &channel_info = m_radio_info.channels_list[chanNum];
 
-            std::unordered_set<std::string> cleared_channels_set;
-            std::unordered_set<std::string> radar_triggered_channels_set;
-
-            s_val = radio->find_child_deep("ChannelMgt.ClearedDfsChannels")->get<std::string>();
-            auto cleared_channels_vec = beerocks::string_utils::str_split(s_val, ',');
-            cleared_channels_set.insert(cleared_channels_vec.cbegin(), cleared_channels_vec.cend());
-
-            s_val = radio->find_child_deep("ChannelMgt.ClearedDfsChannels")->get<std::string>();
-            auto radar_triggered_channels_vec = beerocks::string_utils::str_split(s_val, ',');
-            radar_triggered_channels_set.insert(radar_triggered_channels_vec.cbegin(),
-                                                radar_triggered_channels_vec.cend());
-
             if (son::wireless_utils::is_dfs_channel(chanNum)) {
                 if (cleared_channels_set.find(chan_str) != cleared_channels_set.end()) {
                     channel_info.dfs_state = beerocks::eDfsState::AVAILABLE;
-- 
GitLab

